### PR TITLE
Ecal prefire fix

### PR DESCRIPTION
**Pull request by cverstege**
Currently only corrected valid Jets are used when calculating the prefiring probability. Accordin to the [twiki](https://twiki.cern.ch/twiki/bin/viewauth/CMS/L1PrefiringWeightRecipe), all jets with latest JEC should be used.
Hence, I've added corrected invalid Jets to the product, which can then be used by the `PrefiringWeightProducer` to derive the accurate prefiring weights.

### CODE REVIEW

File: Compile/src/Producers/PrefiringWeightProducer.cc [hunk 1]
```diff
@@ -125,7 +125,14 @@ void PrefiringWeightProducer::Produce(ZJetEvent const& event,
     const auto& thePhotons = product.m_pfPhotons;
 
     //Jets
-    const auto& theJets = product.m_correctedZJets["L1L2L3Res"];
+    auto& validJets = product.m_correctedZJets["L1L2L3Res"];
+    auto& invalidJets = product.m_correctedInvalidJets["L1L2L3Res"];
```
Comment (HerrHorizontal) [inline Compile/src/Producers/PrefiringWeightProducer.cc]: Für den Fall, dass die `L1L2L3Res` Pipeline nicht konfiguriert ist, scheitert das natürlich. Vielleicht sollte man das abfangen.

File: Compile/src/Producers/PrefiringWeightProducer.cc [hunk 2]
```diff
@@ -241,20 +248,20 @@ void PrefiringWeightProducer::Produce(ZJetEvent const& event,
     //Move global prefire weights, as well as those for muons, photons, and jets, to the event
     //Up and Down are switched, since we use the inverse and apply it to data.
     product.m_optionalWeights["prefiringWeight"] =  1./nonPrefiringProba[0];
```
Comment (HerrHorizontal) [inline Compile/src/Producers/PrefiringWeightProducer.cc]: Kann man das nominelle Weight nicht zu dem Event-Weight hinzufügen?

**Review by HerrHorizontal — CHANGES_REQUESTED** [review]
Some minor comments inline.

File: Compile/src/Producers/PrefiringWeightProducer.cc (hunk 2 above)
Comment (cverstege) [reply Compile/src/Producers/PrefiringWeightProducer.cc]: Natürlich

File: Compile/src/Producers/PrefiringWeightProducer.cc (hunk 1 above)
Comment (cverstege) [reply Compile/src/Producers/PrefiringWeightProducer.cc]: Log Warning weiter unten bei 0 Jets

**Comment by cverstege** [comment]
Fixed a bug when calculating the weights. Also moved back to optionalWeights, as those weights are only applied at reco level, not on gen level.
Should be ready now. Please re-review @HerrHorizontal @RHofsaess

**Comment by cverstege** [comment]
I'll squash and merge one approved, so it's only one commit on the master branch in the end.

**Review by HerrHorizontal — APPROVED** [review]
Luckily, you have spotted the bug. Looks good now! :+1:

File: Compile/interface/ZJetProduct.h [hunk 3]
```diff
@@ -181,10 +186,22 @@ class ZJetProduct : public KappaProduct
...
 
-        // Invalid jets, no need for different correction levels
-        return static_cast<KLV*>(m_invalidJets[index]);
+        // Gen jets are always valid
+        if (corrLevel == "Gen") {
+            return 0;
+        }
+        // Uncorrected invalid jet
+        else if (corrLevel == "None") {
+            return static_cast<KLV*>(m_invalidJets[index]);
```
Comment (RHofsaess) [inline Compile/interface/ZJetProduct.h:199]: is this really correct?
eventually this is implicitely casted to usigned long when returning?

**Review by RHofsaess — APPROVED** [review]
Looks good :+1:

**Pull request merged**